File: backend/src/main/java/com/example/mypixel/model/node/CombineNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.model.Parameter;
import com.example.mypixel.model.ParameterType;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.lang.NonNull;

import java.util.HashSet;
import java.util.Map;

@MyPixelNode("Combine")
public class CombineNode extends Node {

    @JsonCreator
    public CombineNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, Parameter> getInputTypes() {
        return Map.of(
                "files_0", Parameter.required(ParameterType.FILEPATH_ARRAY),
                "files_1", Parameter.optional(ParameterType.FILEPATH_ARRAY),
                "files_2", Parameter.optional(ParameterType.FILEPATH_ARRAY),
                "files_3", Parameter.optional(ParameterType.FILEPATH_ARRAY),
                "files_4", Parameter.optional(ParameterType.FILEPATH_ARRAY)
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return Map.of(
                "files_0", new HashSet<String>(),
                "files_1", new HashSet<String>(),
                "files_2", new HashSet<String>(),
                "files_3", new HashSet<String>(),
                "files_4", new HashSet<String>()
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return Map.of("files", Parameter.required(ParameterType.FILEPATH_ARRAY));
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return Map.of(
                "category", "IO",
                "description", "Combine multiple data sources into a single source",
                "color", "#AED581",
                "icon", "CombineIcon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        HashSet<String> files = new HashSet<>();

        for (int i = 0; i < 5; i++) {
            if (inputs.containsKey("files_" + i) && inputs.get("files_" + i) != null) {
                HashSet<String> f = (HashSet<String>) inputs.get("files_" + i);
                files.addAll(f);
            }
        }

        Map<String, Object> outputs;
        outputs = Map.of("files", files);

        return outputs;
    }

    @Override
    public void validate() {

    }
}

File: node/nodes/combine_node.py
from typing import Dict, Any
from node import Node
from storage_client import StorageClient


class CombineNode(Node):

    node_type = "Combine"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files_0": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL",
                "default": set()
            },
            "files_1": {
                "type": "FILEPATH_ARRAY",
                "required": False,
                "widget": "LABEL",
                "default": set()
            },
            "files_2": {
                "type": "FILEPATH_ARRAY",
                "required": False,
                "widget": "LABEL",
                "default": set()
            },
            "files_3": {
                "type": "FILEPATH_ARRAY",
                "required": False,
                "widget": "LABEL",
                "default": set()
            },
            "files_4": {
                "type": "FILEPATH_ARRAY",
                "required": False,
                "widget": "LABEL",
                "default": set()
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL"
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "IO",
            "description": "Combine multiple data sources into a single source",
            "color": "#AED581",
            "icon": "CombineIcon"
        }

    def exec(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        files = set()

        for i in range(5):
            key = f"files_{i}"
            if key in inputs and inputs[key] is not None:
                file_set = inputs[key]
                if not isinstance(file_set, set):
                    file_set = set(file_set) if isinstance(file_set, (list, tuple)) else set()
                files.update(file_set)

        task_id = inputs.get("meta", {}).get("taskId")
        node_id = inputs.get("meta", {}).get("nodeId")

        output_files = []

        for file in files:
            output_files.append(StorageClient.store_from_workspace_to_task(task_id, node_id, file))

        return {"files": output_files}

    def validate(self, inputs: Dict[str, Any]) -> None:
        pass
